### PR TITLE
MODINNREACH-317 ( https://issues.folio.org/browse/MODINREACH-317)

### PR DESCRIPTION
**Pull request by gurleenkaurbp**
MODINNREACH-317 ( https://issues.folio.org/browse/MODINREACH-317)

## Purpose
Clean up "Virtual" Items and Associated Loan, Request, and Inventory Records in case of BORROWING_SITE_CANCEL

## Approach
1. When the borrower performs Cancel Hold action on a transaction or on a request (associated with this transaction) which is in PATRON_HOLD state then with existing behavior another operation will performed of clearing the virtual items for this transaction.
2. In the existing serviceImpl class clearing the virtual items metod will be invoked.
3. Test cases added to test this functionality.
4. Karate test cases updated.
5. Sonar coverage is 100% for the introduction of the new lines.
6. Necessary loggers added.

## Screenshots -
MicrosoftTeams-image (10)
MicrosoftTeams-image (9)
MicrosoftTeams-image (8)
MicrosoftTeams-image (12)
MicrosoftTeams-image (11)

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/ItemServiceImpl.java [hunk 1]
```diff
@@ -37,6 +38,13 @@ public Optional<InventoryItemDTO> find(UUID itemId) {
     return inventoryClient.findItem(itemId);
   }
 
+  @Override
+  public void delete(UUID itemId) {
+      inventoryClient.findItem(itemId).
+        ifPresentOrElse(p -> inventoryClient.deleteItem(itemId),
+          () -> log.info("No item found with itemId:{}",itemId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/ItemServiceImpl.java]: Missing space after comma.

File: src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java [hunk 2]
```diff
@@ -46,6 +46,13 @@ public Optional<LoanDTO> find(UUID loanId) {
     return circulationClient.findLoan(loanId);
   }
 
+  @Override
+  public void delete(UUID loanId) {
+    circulationClient.findLoan(loanId)
+      .ifPresentOrElse(loanDTO -> circulationClient.deleteLoan(loanId)
+        ,()->log.info("Loan not found with loanId:{}",loanId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java]: Missing space after both commas.

File: src/main/java/org/folio/innreach/domain/service/impl/RequestServiceImpl.java [hunk 3]
```diff
@@ -257,6 +257,14 @@ public RequestDTO findRequest(UUID requestId) {
       "No request found with id = " + requestId));
   }
 
+  @Override
+  public void deleteRequest(UUID requestId) {
+    circulationClient.findRequest(requestId)
+      .ifPresentOrElse(requestDTO -> circulationClient.deleteRequest(requestId)
+        ,()-> log.info("Request not present with requestid:"+requestId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/RequestServiceImpl.java]: Probably want space after comma and space before and after `+`.

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java [hunk 4]
```diff
@@ -1807,6 +1807,41 @@ void cancelPatronHold_when_TransactionIsOnHoldOrTransfer_and_RequestIsClosed(Tra
 
     assertNotNull(updatedTransaction);
     assertEquals(TransactionStateEnum.BORROWING_SITE_CANCEL, updatedTransaction.getState());
+//    assertEquals(null,updatedTransaction.getHold().getFolioHoldingId());
+//    assertEquals(null,updatedTransaction.getHold().getFolioLoanId());
+//    assertEquals(null,updatedTransaction.getHold().getFolioInstanceId());
+//    assertEquals(null,updatedTransaction.getHold().getFolioItemId());
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java]: Should these commented out items maybe be removed?

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java [hunk 5]
```diff
@@ -1807,6 +1807,41 @@ void cancelPatronHold_when_TransactionIsOnHoldOrTransfer_and_RequestIsClosed(Tra
...
+
+    assertEquals(HttpStatus.OK, responseEntity.getStatusCode());
+    var updatedTransaction = responseEntity.getBody();
+
+    assertNotNull(updatedTransaction);
+    assertEquals(TransactionStateEnum.BORROWING_SITE_CANCEL, updatedTransaction.getState());
+//    assertEquals(null,updatedTransaction.getHold().getFolioHoldingId());
+//    assertEquals(null,updatedTransaction.getHold().getFolioLoanId());
+//    assertEquals(null,updatedTransaction.getHold().getFolioInstanceId());
+//    assertEquals(null,updatedTransaction.getHold().getFolioItemId());
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java]: Same as above.

File: src/test/java/org/folio/innreach/controller/d2ir/CirculationApiTest.java [hunk 6]
```diff
@@ -241,10 +244,15 @@ void patronHold_cancelRequest() throws Exception {
...
     stubGet(format("%s/%s", ITEMS_URL, PRE_POPULATED_ITEM_ID), "inventory/item-response.json");
     stubPut(format("%s/%s", ITEMS_URL, PRE_POPULATED_ITEM_ID));
+    stubDelete(format("%s/%s", ITEMS_URL, PRE_POPULATED_ITEM_ID));
     stubGet(format("%s/%s", HOLDINGS_URL, NEW_HOLDING_ID), "inventory-storage/holding-response.json");
     stubPut(format("%s/%s", HOLDINGS_URL, NEW_HOLDING_ID));
+    stubDelete(format("%s/%s", HOLDINGS_URL, FOLIO_HOLDING_ID));
     stubGet(format("%s/%s", REQUESTS_URL, PRE_POPULATED_REQUEST_ID), "circulation/item-request-response.json");
     stubPut(format("%s/%s", REQUESTS_URL, PRE_POPULATED_REQUEST_ID));
+    stubDelete(format("%s/%s", INSTANCES_URL, FOLIO_INSTANCE_ID));
+    stubGet(LOAN_URL,"circulation/loan-response.json");
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/d2ir/CirculationApiTest.java]: Space after comma

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 7]
```diff
@@ -46,6 +46,7 @@
 
 import lombok.RequiredArgsConstructor;
 import lombok.extern.log4j.Log4j2;
+import org.folio.innreach.domain.service.*;
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Usually wildcarding of imports is discouraged.

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 8]
```diff
@@ -30,9 +30,7 @@
 import static org.folio.innreach.dto.ItemStatus.NameEnum.PAGED;
 import static org.folio.innreach.util.DateHelper.toEpochSec;
 import static org.folio.innreach.util.DateHelper.toInstantTruncatedToSec;
-import static org.folio.innreach.util.InnReachTransactionUtils.clearCentralPatronInfo;
-import static org.folio.innreach.util.InnReachTransactionUtils.clearPatronAndItemInfo;
-import static org.folio.innreach.util.InnReachTransactionUtils.verifyState;
+import static org.folio.innreach.util.InnReachTransactionUtils.*;
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Same as above.

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java (hunk 4 above)
Comment (gurleenkaurbp) [reply src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java]: These can be un-commented now, I have made the require change and committed back.

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java (hunk 5 above)
Comment (gurleenkaurbp) [reply src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java]: These can be un-commented now, I have made the require change and committed back.

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 8 above)
Comment (gurleenkaurbp) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Done

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 9]
```diff
@@ -30,9 +30,8 @@
...
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Can this wildcard also be replaced with only the needed imports?

**Review by Arghya22 — APPROVED** [review]

**Review by nirmalr-shah — APPROVED** [review]

**Pull request merged**